src/object/cli_obj.c

```diff
@@ -6123,6 +6197,115 @@ dc_obj_list_rec(tse_task_t *task)
 	return obj_list_common(task, DAOS_OBJ_RECX_RPC_ENUMERATE, args);
 }
 
+static int
+shard_k2a_prep(struct shard_auxi_args *shard_auxi, struct dc_object *obj,
+	       struct obj_auxi_args *obj_auxi, uint32_t grp_idx)
+{
+	daos_obj_key2anchor_t	*obj_args;
+	struct shard_k2a_args	*shard_arg;
+
+	obj_args = dc_task_get_args(obj_auxi->obj_task);
+	shard_arg = container_of(shard_auxi, struct shard_k2a_args, ka_auxi);
+	shard_arg->ka_anchor = obj_args->anchor;
+
+	return 0;
+}
+
+int
+dc_obj_key2anchor(tse_task_t *task)
+{
+	daos_obj_key2anchor_t	*args;
+	struct obj_auxi_args	*obj_auxi;
+	struct dc_object	*obj;
+	unsigned int		map_ver = 0;
+	struct dtx_epoch	epoch;
+	uint32_t		shard;
+	uint32_t		shard_cnt;
+	int			grp_idx = 0;
+	int			rc = 0;
+
+	args = dc_task_get_args(task);
+	D_ASSERTF(args != NULL, "Task Argument OPC does not match DC OPC\n");
+
+	if (args->anchor == NULL) {
+		D_ERROR("Invalid anchor to daos_obj_key2anchor\n");
+		D_GOTO(err_task, rc);
+	}
+
+	rc = obj_req_valid(task, args, DAOS_OBJ_RPC_KEY2ANCHOR, &epoch, &map_ver, &obj);
+	if (rc != 0)
+		D_GOTO(err_task, rc);
+
+	rc = obj_task_init(task, DAOS_OBJ_RPC_KEY2ANCHOR, map_ver, args->th, &obj_auxi, obj);
+	if (rc != 0)
+		D_GOTO(err_obj, rc);
+
+	obj_auxi->dkey_hash = obj_dkey2hash(obj->cob_md.omd_id, args->dkey);
+	grp_idx = obj_dkey2grpidx(obj, obj_auxi->dkey_hash, map_ver);
+	if (grp_idx < 0) {
+		D_ERROR(DF_OID" can not find grp %d\n", DP_OID(obj->cob_md.omd_id), grp_idx);
+		D_GOTO(err_obj, rc = grp_idx);
+	}
+
+	if (obj_auxi->is_ec_obj) {
+		struct shard_anchors	*sub_anchors = NULL;
+
+		rc = obj_ec_get_parity_or_alldata_shard(obj_auxi, map_ver, grp_idx, args->dkey,
+							&shard_cnt);
+		shard_cnt = obj_get_grp_size(obj);
```

Comment:
probably it should use the shard_cnt got from obj_ec_get_parity_or_alldata_shard(), in case parity rotation is not enabled (shard_cnt = 1), and also obj_get_grp_size() will also include the extending shard during online OSA, which is not needed for enumeration.

Reply:
well i think shard_cnt was always returned as 1 so i had to do this, but i will check again to make sure.

Reply:
ok i confirmed that this is required as obj_ec_get_parity_or_alldata_shard always returning 1 for shard_cnt.. since i assume dkey is not NULL. so to be able to go query all shards, we have to do this.

Reply:
ah, hmm, probably should be

if (obj_ec_parity_rotate_enable())
         shard_cnt = daos_oclass_grp_size();